### PR TITLE
Fix bulk load storage slots

### LINKED ISSUE #11551: Bulk slot loading returns inconsistent response
### Description

Customers have complained that the new bulk slot loading and caching mechanism added in 0.133 causes results to the same contract with different data to return inconsistent responses.

### Steps to reproduce

1. Enable bulk slot loading
2. Run the following 4 calls in sequence
```json
{
    "to": "0x00000000000000000000000000000000003c4196",
    "data": "0x351fb47800000000000000000000000027784a45fdee4d8f355237a4e92f259c1d7d7617fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe"
}
{
    "data": "0x351fb47800000000000000000000000027784a45fdee4d8f355237a4e92f259c1d7d7617ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
    "to": "0x00000000000000000000000000000000003c4196"
}
{
    "data": "0x351fb47800000000000000000000000027784a45fdee4d8f355237a4e92f259c1d7d76170000000000000000000000000000000000000000000000000000000000000000",
    "to": "0x00000000000000000000000000000000003c4196"
}
{
    "data": "0x351fb47800000000000000000000000027784a45fdee4d8f355237a4e92f259c1d7d76170000000000000000000000000000000000000000000000000000000000000001",
    "to": "0x00000000000000000000000000000000003c4196"
}
```
Only the middle two (-1 and 0) should return data, the others will return a 0 array.

### Additional context

_No response_

### Hedera network

mainnet

### Version

v0.133.0

### Operating system

None

### PR DESCRIPTION
**Pull request by kselveliev**
**Description**:
Fix bulk load storage slots.

Added a couple of integration tests that test concurrent calls to cover specific corner cases and also eviction corner case.

Fixes #11551

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractStateServiceTest.java [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 package org.hiero.mirror.web3.service;
 
+import static net.i2p.crypto.eddsa.Utils.hexToBytes;
```
Comment (bilyana-gospodinova) [inline web3/src/test/java/org/hiero/mirror/web3/service/ContractStateServiceTest.java]: We can probably use some other import like `org.apache.commons.codec.binary.Hex.decodeHex(String)` similar to other classes that we have.

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java [hunk 2]
```diff
@@ -83,20 +83,17 @@ public Optional<byte[]> findStorageByBlockTimestamp(
...
+        final var contractSlotsCache = ((CaffeineCache) this.contractSlotsCache.get(
+                contractId, () -> cacheManagerSlotsPerContract.getCache(contractId.toString())));
         // Cached slot keys for contract, whose slot values are not present in the contractStateCache
         final var cachedSlots = new ArrayList<byte[]>();
-        ((CaffeineCache) contractSlotsCache).getNativeCache().asMap().keySet().forEach(slot -> {
-            final var slotBytes = ((ByteBuffer) slot).array();
-            final var value = contractStateCache.putIfAbsent(generateCacheKey(contractId, slotBytes), EMPTY_VALUE);
-
-            if (value == null) {
+        synchronized (contractSlotsCache) {
```
Comment (kselveliev) [inline web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Proves needed to avoid corner case when eviction is being done.

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java [hunk 3]
```diff
@@ -83,20 +83,17 @@ public Optional<byte[]> findStorageByBlockTimestamp(
...
-                contractId, () -> cacheManagerSlotsPerContract.getCache(contractId.toString()));
-        contractSlotsCache.putIfAbsent(ByteBuffer.wrap(key), EMPTY_VALUE);
-
+        final var contractSlotsCache = ((CaffeineCache) this.contractSlotsCache.get(
+                contractId, () -> cacheManagerSlotsPerContract.getCache(contractId.toString())));
         // Cached slot keys for contract, whose slot values are not present in the contractStateCache
         final var cachedSlots = new ArrayList<byte[]>();
-        ((CaffeineCache) contractSlotsCache).getNativeCache().asMap().keySet().forEach(slot -> {
-            final var slotBytes = ((ByteBuffer) slot).array();
-            final var value = contractStateCache.putIfAbsent(generateCacheKey(contractId, slotBytes), EMPTY_VALUE);
```
Comment (kselveliev) [inline web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java:94]: This is unneeded, now removed.
Puts empty values in contract state cache and does pretty bad in concurrent environment.

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java [hunk 4]
```diff
@@ -105,7 +102,6 @@ private Optional<byte[]> findStorageBatch(final EntityId contractId, final byte[
             final byte[] slotKey = contractSlotValue.getSlot();
             final byte[] slotValue = contractSlotValue.getValue();
             contractStateCache.put(generateCacheKey(contractId, slotKey), slotValue);
-            contractSlotsCache.put(ByteBuffer.wrap(slotKey), EMPTY_VALUE);
```
Comment (kselveliev) [inline web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java:108]: Also unneeded.

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java (hunk 2 above)
Comment (sdimitrov9) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Consider removing the synchronized block and modify the return statement of the method to:

```
return Optional.ofNullable(cachedValue)
                .or(() -> contractStateRepository.findStorage(contractId.getId(), key));
```

This way we should be able to avoid the problem cause by the corner case where the cache eviction clears the key that was just added by the `contractSlotsCache.putIfAbsent(ByteBuffer.wrap(key), EMPTY_VALUE);`

Comment (sdimitrov9) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: The above approach would deal with the corner case, but it will do an unnecessary query in cases were the slot key is not present in the db at all, so it will impact performance. Another approach should be investigated.

Comment (sdimitrov9) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Consider adding         
`boolean evicted = !cachedSlots.contains(ByteBuffer.wrap(key).array());`
after the findStorageBatch statement and the following if statement before the final return:
```
if (cachedValue == null && evicted) {
            return contractStateRepository.findStorage(contractId.getId(), key);
        }
```

Comment (kselveliev) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Addressed

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java [hunk 5]
```diff
@@ -83,35 +83,35 @@ public Optional<byte[]> findStorageByBlockTimestamp(
...
 
             if (Arrays.equals(slotKey, key)) {
                 cachedValue = slotValue;
             }
         }
 
+        // If the cache key was evicted and hasn't been requested since, the cached value will be null.
+        // In that case, fall back to the original query.
+        if (cachedValue == null && evicted) {
```
Comment (sdimitrov9) [inline web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: I think If slot key is evicted cachedValue will always be null. In that case the if condition should be just:
`if (evicted) { ....}`
Considering the complexity of the code it might be reasonable to rename evicted to isKeyEvictedFromCache for better clarity.

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractStateServiceTest.java (hunk 1 above)
Comment (kselveliev) [reply web3/src/test/java/org/hiero/mirror/web3/service/ContractStateServiceTest.java]: Addressed

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java (hunk 5 above)
Comment (kselveliev) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Correct, done

**Review by sdimitrov9 — APPROVED** [review]

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java [hunk 6]
```diff
@@ -83,35 +83,36 @@ public Optional<byte[]> findStorageByBlockTimestamp(
...
-        final var contractSlotsCache = this.contractSlotsCache.get(
-                contractId, () -> cacheManagerSlotsPerContract.getCache(contractId.toString()));
-        contractSlotsCache.putIfAbsent(ByteBuffer.wrap(key), EMPTY_VALUE);
-
+        final var contractSlotsCache = ((CaffeineCache) this.contractSlotsCache.get(
+                contractId, () -> cacheManagerSlotsPerContract.getCache(contractId.toString())));
         // Cached slot keys for contract, whose slot values are not present in the contractStateCache
         final var cachedSlots = new ArrayList<byte[]>();
-        ((CaffeineCache) contractSlotsCache).getNativeCache().asMap().keySet().forEach(slot -> {
+        contractSlotsCache.putIfAbsent(ByteBuffer.wrap(key), EMPTY_VALUE);
```
Comment (xin-hedera) [inline web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: `cachedSlots` has to be a `HashSet<ByteBuffer>` for `contains` to work, also for efficient `contains` check

```suggestion
        final var wrappedkey = ByteBuffer.wrap(key);
        contractSlotsCache.putIfAbsent(wrappedkey, EMPTY_VALUE);
        final var cachedSlots = contractSlotsCache.getNativeCache().asMap().keySet();
        ...
        
        boolean isKeyEvictedFromCache = !cachedSlots.contains(wrappedKey);
```

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java [hunk 7]
```diff
@@ -83,35 +83,36 @@ public Optional<byte[]> findStorageByBlockTimestamp(
...
 
             if (Arrays.equals(slotKey, key)) {
                 cachedValue = slotValue;
             }
         }
 
+        // If the cache key was evicted and hasn't been requested since, the cached value will be null.
+        // In that case, fall back to the original query.
+        if (isKeyEvictedFromCache) {
+            return contractStateRepository.findStorage(contractId.getId(), key);
```
Comment (xin-hedera) [inline web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java:119]: cache the slot value

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java [hunk 8]
```diff
@@ -83,35 +83,36 @@ public Optional<byte[]> findStorageByBlockTimestamp(
...
-                contractId, () -> cacheManagerSlotsPerContract.getCache(contractId.toString()));
-        contractSlotsCache.putIfAbsent(ByteBuffer.wrap(key), EMPTY_VALUE);
-
+        final var contractSlotsCache = ((CaffeineCache) this.contractSlotsCache.get(
+                contractId, () -> cacheManagerSlotsPerContract.getCache(contractId.toString())));
         // Cached slot keys for contract, whose slot values are not present in the contractStateCache
         final var cachedSlots = new ArrayList<byte[]>();
-        ((CaffeineCache) contractSlotsCache).getNativeCache().asMap().keySet().forEach(slot -> {
+        contractSlotsCache.putIfAbsent(ByteBuffer.wrap(key), EMPTY_VALUE);
+        (contractSlotsCache).getNativeCache().asMap().keySet().forEach(slot -> {
```
Comment (steven-sheehy) [inline web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Unnecessary parenthesis
```suggestion
        contractSlotsCache.getNativeCache().asMap().keySet().forEach(slot -> {
```

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java (hunk 7 above)
Comment (steven-sheehy) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java:119]: The repository has `@Cacheable` so this should already be cached before returning, unless there's some bug

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java (hunk 6 above)
Comment (steven-sheehy) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Maybe instead we can just compare the key inside the first loop to see if it's present?

Comment (sdimitrov9) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: I think If we do the check in the first loop the performance won’t change, considering the Caffeine cache is using ConcurrentHashMap as an underlying structure. Since it is unordered the check via iteration would has O(n/2) complexity on average similar to the ArrayList contains(). Also the lambda expression would become more complex.
On the other hand,  the contains() method  of the HashSet would have constant time complexity on average compared to O(n) in the ArrayList case.

Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: The `contains` in ConcurrentHashMap should be also constant.

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java (hunk 8 above)
Comment (kselveliev) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Addressed

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java (hunk 7 above)
Comment (xin-hedera) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java:119]: missed that

File: web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java (hunk 6 above)
Comment (sdimitrov9) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: It is, but iterating through the structure isn't.

Comment (xin-hedera) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Agree we can also do the `isKeyEvictedFromCache` check inside the loop. We can't avoid the loop because `contractStateRepository.findStorageBatch` takes `List<byte[]>` while the keyset in the underlying `cache.asMap()` is a `Set<ByteBuffer>`.

Comment (kselveliev) [reply web3/src/main/java/org/hiero/mirror/web3/service/ContractStateServiceImpl.java]: Refactored the code to work with the approach that Steven suggested initially.
Comparing the key inside the loop to check if it's present.

**Review by xin-hedera — APPROVED** [review]
LGTM
